pkg/sentry/fsimpl/fuse/regular_file.go

```diff
@@ -49,149 +49,77 @@ func (fd *regularFileFD) PRead(ctx context.Context, dst usermem.IOSequence, offs
-
-	// bytes to read.
-	size uint32
-	// offset of read.
-	off uint64
-
-	// actual bytes read.
-	n uint32
-	// read error.
-	err error
-
-	// buffer for bytes read,
-	// ideally it shares the same array with the slice in FUSE response
-	// for the reads that can fit in one FUSE_READ request.
-	buf []byte
-}
-
-func (rw *regularFDReadWriter) fs() *filesystem {
-	return rw.fd.inode().fs
-}
-
-var regularFdReadWriterPool = sync.Pool{
-	New: func() interface{} {
-		return &regularFDReadWriter{}
-	},
-}
-
-func getRegularFDReadWriter(ctx context.Context, fd *regularFileFD, size uint32, offset int64) *regularFDReadWriter {
-	rw := regularFdReadWriterPool.Get().(*regularFDReadWriter)
-	rw.ctx = ctx
-	rw.fd = fd
-	rw.size = size
-	rw.off = uint64(offset)
-	return rw
-}
-
-func putRegularFDReadWriter(rw *regularFDReadWriter) {
-	rw.ctx = nil
-	rw.fd = nil
-	rw.buf = nil
-	rw.n = 0
-	rw.err = nil
-	regularFdReadWriterPool.Put(rw)
-}
-
-// read handles and issues the actual FUSE read request.
-// See ReadToBlocks() regarding its purpose.
-func (rw *regularFDReadWriter) read() {
-	// TODO(gvisor.dev/issue/3237): support indirect IO (e.g. caching):
-	// use caching when possible.
-
-	inode := rw.fd.inode()
+	inode := fd.inode()
 
 	// Reading beyond EOF, update file size if outdated.
-	if rw.off+uint64(rw.size) >= atomic.LoadUint64(&inode.size) {
-		if err := inode.reviseAttr(rw.ctx); err != nil {
-			rw.err = err
-			return
+	if uint64(offset)+size > atomic.LoadUint64(&inode.size) {
+		if err := inode.reviseAttr(ctx); err != nil {
+			return 0, err
 		}
 		// If the offset after update is still too large, return error.
-		if rw.off >= atomic.LoadUint64(&inode.size) {
-			rw.err = io.EOF
-			return
+		if uint64(offset) >= atomic.LoadUint64(&inode.size) {
+			return 0, io.EOF
 		}
 	}
 
 	// Truncate the read with updated file size.
 	fileSize := atomic.LoadUint64(&inode.size)
-	if rw.off+uint64(rw.size) > fileSize {
-		// This uint32 conversion will not overflow.
-		// Since rw.off < fileSize and the difference
-		// between must be less than rw.size to make
-		// the if condition true.
-		rw.size = uint32(fileSize - rw.off)
-	}
-
-	// Send the FUSE_READ request and store the data in rw.
-	rw.buf, rw.n, rw.err = rw.fs().ReadInPages(rw.ctx, rw.fd, rw.off, rw.size)
-}
-
-// ReadToBlocks implements safemem.Reader.ReadToBlocks.
-// Due to a deadlock (both the caller of ReadToBlocks and the kernelTask.Block()
-// will try to acquire the same lock), have to separate the rw.read() from the
-// ReadToBlocks() function. Therefore, ReadToBlocks() only handles copying
-// the result into user memory while read() handles the actual reading.
-func (rw *regularFDReadWriter) ReadToBlocks(dsts safemem.BlockSeq) (uint64, error) {
-	if rw.err != nil {
-		return 0, rw.err
+	if uint64(offset)+size > fileSize {
+		size = fileSize - uint64(offset)
 	}
 
-	if dsts.IsEmpty() {
-		return 0, nil
+	buffers, n, err := inode.fs.ReadInPages(ctx, fd, uint64(offset), uint32(size))
```

Comment:
Can `n` be returned as `uint64` type?

Reply:
It might be better to keep it the same type as `size` of FUSE struct. The caller should not be a reason for the callee to change the abstraction